Hda example with distillation

radhakrishnatg opened the pull request:
## Fixes # .


## Proposed changes:
- Adding the example from the IDAES paper that includes a distillation column and kinetic reactor
- Includes the reaction package with Arrhenius factor and activation energy
- Square problem + optimization (OUU will be added as a separate notebook)

### Legal Acknowledgement

By contributing to this software project, I agree to the following terms and conditions for my contribution:

1. I agree my contributions are submitted under the license terms described in the LICENSE.txt file at the top level of this directory.
2. I represent I am authorized to make the contributions and grant the license. If my employer has rights to intellectual property that includes these contributions, I represent that I have received permission to make contributions and grant the required license on behalf of that employer.

andrewlee94 reviewed (changes requested):
It looks like the tests are failing on the old version of the HDA example due to the change in the units used for the property package.

We should remove the old version (or at least move it to a deprecated folder somewhere) to avoid any confusion for users.

andrewlee94 commented:
Whilst we are at it, should we try to switch to the newer generic property framework as well (although that should maybe be a separate PR).

andrewlee94 commented:
Do we really want to keep 2 versions of this example around? My feeling is it would be better to have a single example, both for ease of maintenance and to avoid confusion for readers.

radhakrishnatg commented:
@andrewlee94 Agree, since there is only a small difference, it would make sense to keep only one version. Should I remove the old version? If you prefer that I move it to a different folder, could you please suggest which one?

jghouse88 commented:
@radhakrishnatg The update make the tests pass but the doc build is failing. @dangunter any pointers?

jghouse88 commented:
@andrewlee94 I would like to have the 2 flash version as well. It will serve well in workshops.

andrewlee94 commented:
@jghouse88 If we are going to keep both, we need to make sure we have guidance to the user on what the difference is and why they should look at each one. I would also suggest putting one of them in a separate subfolder somewhere to keep the top level folder cleaner of new users (maybe the Flash version, as it is a special case for in person workshops where there is limited time?).

radhakrishnatg commented:
@andrewlee94 At the beginning of the notebook, I have added the following note.
"This tutorial will be similar to the one in HDA_flowsheet_exercise.ipynb, except that we use a distillation column instead of a second flash (F102) to produce benzene and toluene products."

jghouse88 reviewed (approved):
@radhakrishnatg this looks really good. I tested the files and everything runs. I however noticed that most of the links in the `src/notebook_index.ipynb` are not working. This is not something coming from this PR but we need to look at this. 

The only link that worked for me was `data_management_framework ` - no surprises there as that was done by @dangunter and he is obviously something correct that the rest of us are not doing.

andrewlee94 commented on src/Examples/Flowsheets/HDA_flowsheet_with_distillation_solution_testing.ipynb:
```diff
@@ -0,0 +1,1773 @@
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "## Connecting Unit Models using Arcs\n",
```
It is probably worth reminding the user of the issue with the SD Tool here and hence that we will add the distillation column later.

radhakrishnatg replied:
I have mentioned this just a few lines above, so I felt it would be a repetition. Please let me know if you still want me to add it.

andrewlee94 commented on src/Examples/Flowsheets/HDA_flowsheet_with_distillation_solution_testing.ipynb:
```diff
@@ -0,0 +1,1773 @@
+    "seq = SequentialDecomposition()\n",
+    "seq.options.select_tear_method = \"heuristic\"\n",
+    "seq.options.tear_method = \"Wegstein\"\n",
+    "seq.options.iterLim = 5\n",
```
Seeing as we don't need to fully converge the flowsheet, I would suggest decreasing this to 3.

andrewlee94 commented on src/Examples/Flowsheets/HDA_flowsheet_with_distillation_solution_testing.ipynb:
```diff
@@ -0,0 +1,1773 @@
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "The SequentialDecomposition tool has determined that the tear stream is the mixer outlet (s03 in the Figure above). We will need to provide a reasonable guess for this."
```
I think we should add an explanation of how we got the guess too - it is the sum of the two feed streams (i.e. assuming negligible recycle for the first pass).

radhakrishnatg replied:
Added the following explanation:
"For the initial guess, we assume that the flowrate of the recycle stream (s09) is zero. Consequently, the flow rate of the stream s03 is simply the sum of the flowrates of the toluene feed and hydrogen feed streams. Further, since the temperature and the pressure of both the toluene and hydrogen feed streams are the same, we specify their values as the initial guess for the temperature and pressure of the stream s03."

andrewlee94 commented on src/Examples/Flowsheets/HDA_flowsheet_with_distillation_solution_testing.ipynb:
```diff
@@ -0,0 +1,1773 @@
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "We are now ready to initialize our flowsheet in a sequential mode. Note that we specifically set the iteration limit to be 5 as we are trying to use this tool only to get a good set of initial values such that IPOPT can then take over and solve this flowsheet for us. "
```
Remember to update this is you change the number of iterations.

andrewlee94 commented on src/Examples/Flowsheets/HDA_flowsheet_with_distillation_solution_testing.ipynb:
```diff
@@ -0,0 +1,1773 @@
+   "outputs": [],
+   "source": [
+    "# Create the solver object\n",
+    "solver = SolverFactory('ipopt')\n",
```
1. As before, should we change to using `get_solver`?
2. 5000 iterations is way too many - 100 is more than enough and I would even suggest 50.

radhakrishnatg replied:
Switched to get_solver

andrewlee94 commented on src/Examples/Flowsheets/HDA_flowsheet_with_distillation_solution_testing.ipynb:
```diff
@@ -0,0 +1,1773 @@
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "# Check solver solve status\n",
```
Should this be tagged as a test cell?

radhakrishnatg replied:
Yes, it should be. Sorry, I forgot to tag it.

andrewlee94 commented on src/Examples/Flowsheets/HDA_flowsheet_with_distillation_solution_testing.ipynb:
```diff
@@ -0,0 +1,1773 @@
+   "source": [
+    "### Add distillation column \n",
+    "\n",
+    "This completes the simulation of the flowsheet constructed earlier. We will now \n",
```
I would add a little more explanation here, mainly a call back to earlier and say we will now add the distillation column. You can then go into the the specific steps required.

radhakrishnatg replied:
I have modified the text to the following:
"As mentioned earlier, the SequentialDecomposition tool currently does not support the distillation column model. Thus, we have not included the distillation column in the flowsheet. Now that we have a converged flowsheet, we will add the distillation column and simulate the entire flowsheet. In the following, we will..."

Please let me know if you want me to include any other point.

andrewlee94 commented on src/Examples/Flowsheets/HDA_flowsheet_with_distillation_solution_testing.ipynb:
```diff
@@ -0,0 +1,1773 @@
+    "                            TemperatureSpec.atBubblePoint,\n",
+    "                        \"property_package\": m.fs.bt_properties,\n",
+    "                        \"has_heat_transfer\": False,\n",
+    "                        \"has_pressure_change\": False})\n",
```
This is not something for this PR, but I see a lot of Pyomo warnings about implicit replacement of components. We should address that in the main distillation code.

radhakrishnatg replied:
Yeah, Jaffer took a note of it. He said that he would take a look at the distillation code and figure out what the issue is.

andrewlee94 commented on src/Examples/Flowsheets/HDA_flowsheet_with_distillation_solution_testing.ipynb:
```diff
@@ -0,0 +1,1773 @@
+   "source": [
+    "## Analyze the Results of the Square Problem\n",
+    "\n",
+    "How much is the total cost (operating_cost + capital_cost), operating_cost, capital_cost, benzene purity in the distillate from the distilation column, and conversion of toluene in the reactor?"
```
I would remove the underscores from the names here - you are describing the quantities rather than using the actual variable names,.

andrewlee94 commented on src/Examples/Flowsheets/HDA_flowsheet_with_distillation_solution_testing.ipynb:
```diff
@@ -0,0 +1,1773 @@
+    "print('R101 outlet temperature = ', value(m.fs.R101.outlet.temperature[0]), 'K')\n",
+    "\n",
+    "print()\n",
+    "print('F101 outlet temperature = ', value(m.fs.F101.vap_outlet.temperature[0]), 'K')"
```
I think we should add a brief discussion/explanation of the results. Things we should note:

* improved product purity and less overhead loss
* decrease in operating cost but increase in capital cost
* observations on key operating conditions (notably any temperatures on bounds).

radhakrishnatg replied:
I have added a brief discussion, similar to the one in the paper.
"### Key Takeaways

Observe that optimization was able to reduce the yearly operating cost from \\$427,593 to \\$408,342 (~4.5%). However, the amortized capital cost more than doubled from \\$14,704 to \\$29,927 due to the need to increase the conversion in the reactor (from 75% to 93%) to meet the production and purity constraints. 

Further, observe that the product flow rate and product purity are at their minimum values (0.18 mol/s and 99%, respectively). This is expected as increasing recovery would require more energy and cost to purify the product.


Finally, observe that the operating temperature of the flash (F101) is almost at its lower bound. This helps in minimizing the amount of benzene in the vapor stream leaving the flash."

andrewlee94 reviewed (changes requested):
Looking good. I had a lot of comments but most are small things.

radhakrishnatg commented on src/Examples/Flowsheets/hda_ideal_VLE.py:
```diff
@@ -0,0 +1,1025 @@
+##############################################################################
```
We can use the file in the Tutorials section. I put a duplicate here hoping that it would be easier to access the file. Please let me know if you prefer using the one in the Tutorials section, I will remove this file.

andrewlee94 replied:
Duplicating files is never a good idea - always use an existing file if there is one.

radhakrishnatg commented on src/Examples/Flowsheets/HDA_flowsheet_with_distillation_solution_testing.ipynb:
```diff
@@ -0,0 +1,1773 @@
+    "\n",
+    "## Note\n",
+    "\n",
+    "This tutorial will be similar to the one in HDA_flowsheet_exercise.ipynb, except that we use a distillation column instead of a second flash (F102) to produce benzene and toluene products.\n",
```
I have modified the text. It now reads: "This tutorial will be similar to the HDA flowsheet tutorial in the Tutorials section, except ..."

radhakrishnatg commented on src/Examples/Flowsheets/HDA_flowsheet_with_distillation_solution_testing.ipynb:
```diff
@@ -0,0 +1,1773 @@
+    "present by Douglas, J.M., Chemical  Design of Chemical Processes, 1988,\n",
+    "McGraw-Hill.\n",
+    "\n",
+    "The flowsheet that we will be using for this module is shown below with the stream conditions. We will be processing toluene and hydrogen to produce at least 370 TPY of benzene. As shown in the flowsheet, we use a flash tank, F101, to separate out the non-condensibles, and a distillation column, D101, to further separate the benzene-toluene mixture to improve the benzene purity.  The non-con ...
```
I have added the benzene-toluene package in the description, and pointed out that we will be using two property packages. 
"We will be using two thermodynamic packages: one (first in the list above) containing all four components (i.e., toluene, hydrogen, benzene, and methane) and the other (second in the list above) containing benzene and toluene only. The latter is needed to simplify the VLE calculations in the distillation column model."

radhakrishnatg commented on src/Examples/Flowsheets/HDA_flowsheet_with_distillation_solution_testing.ipynb:
```diff
@@ -0,0 +1,1773 @@
+    "\n",
+    "The flowsheet that we will be using for this module is shown below with the stream conditions. We will be processing toluene and hydrogen to produce at least 370 TPY of benzene. As shown in the flowsheet, we use a flash tank, F101, to separate out the non-condensibles, and a distillation column, D101, to further separate the benzene-toluene mixture to improve the benzene purity.  The non-con ...
+    "\n",
+    "- `hda_ideal_VLE.py`\n",
```
Included the package in the description.

radhakrishnatg replied:
I have moved the description about the translator block to the beginning of the notebook (below the flowsheet).

radhakrishnatg commented on src/Examples/Flowsheets/HDA_flowsheet_with_distillation_solution_testing.ipynb:
```diff
@@ -0,0 +1,1773 @@
+    "- `hda_ideal_VLE.py`\n",
+    "- `hda_reaction_kinetic.py`\n",
+    "\n",
+    "The state variables chosen for the property package are **flows of component by phase, temperature and pressure**. The components considered are: **toluene, hydrogen, benzene and methane**. Therefore, every stream has 8 flow variables, 1 temperature and 1 pressure variable. \n",
```
I have removed the text in the revised version.

radhakrishnatg commented on src/Examples/Flowsheets/HDA_flowsheet_with_distillation_solution_testing.ipynb:
```diff
@@ -0,0 +1,1773 @@
+    "- StoichiometricReactor\n",
+    "- Flash\n",
+    "- Separator (splitter) \n",
+    "- PressureChanger"
```
I have added them, and included a brief description.
- Translator (to switch from one property package to another)
- TrayColumn (distillation column)
- CondenserType (Type of the overhead condenser: complete or partial)
- TemperatureSpec (Temperature specification inside the condenser)

radhakrishnatg commented on src/Examples/Flowsheets/HDA_flowsheet_with_distillation_solution_testing.ipynb:
```diff
@@ -0,0 +1,1773 @@
+   "source": [
+    "## Constructing the Flowsheet\n",
+    "\n",
+    "We have now imported all the components, unit models, and property modules we need to construct a flowsheet. Let us create a ConcreteModel and add the flowsheet block as we did in the earlier module. "
```
Modified the sentence to "Let us create a ConcreteModel and add the flowsheet block to it."

radhakrishnatg commented on src/Examples/Flowsheets/HDA_flowsheet_with_distillation_solution_testing.ipynb:
```diff
@@ -0,0 +1,1773 @@
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "We now need to add the property packages to the flowsheet. Unlike Module 1, where we only had a thermo property package, for this flowsheet we will also need to add a reaction property package. "
```
Modified the sentence to "We will now add the thermophysical and reaction packages to the flowsheet." I have replaced word "thermo" with "thermophysical" everywhere in the notebook.

radhakrishnatg commented on src/Examples/Flowsheets/HDA_flowsheet_with_distillation_solution_testing.ipynb:
```diff
@@ -0,0 +1,1773 @@
+   "outputs": [],
+   "source": [
+    "# Property package for benzene, toluene, hydrogen, methane mixture\n",
+    "m.fs.thermo_params = HDAParameterBlock()\n",
```
I have changed the names to "BTHM_params" and "BT_params".

radhakrishnatg commented on src/Examples/Flowsheets/HDA_flowsheet_with_distillation_solution_testing.ipynb:
```diff
@@ -0,0 +1,1773 @@
+   "source": [
+    "## Adding Unit Models\n",
+    "\n",
+    "Let us start adding the unit models we have imported to the flowsheet. Here, we are adding the Mixer (assigned a name M101) and a Heater (assigned a name H101). Note that, all unit models need to be given a property package argument. In addition to that, there are several arguments depending on the unit model, please refer to the documentation for more details (https://idaes-pse.readthedocs. ...
```
I have modified the text. It now reads
"Let us start adding the unit models we have imported to the flowsheet. Here, we are adding the Mixer (assigned a name M101) and a Heater (assigned a name H101). Note that, all unit models need to be given a property package argument. In addition, the Mixer unit model needs a list consisting of the inlets (toluene feed, hydrogen feed and vapor recycle streams in this case)."
Please let me know if I need to condense it further.

radhakrishnatg commented on src/Examples/Flowsheets/HDA_flowsheet_with_distillation_solution_testing.ipynb:
```diff
@@ -0,0 +1,1773 @@
+    "\n",
+    "# Adding the heater H101 to the flowsheet\n",
+    "m.fs.H101 = Heater(default={\"property_package\": m.fs.thermo_params,\n",
+    "                            \"has_pressure_change\": False,\n",
```
Thank you for the tip! I have now removed the options which are False.

radhakrishnatg commented on src/Examples/Flowsheets/HDA_flowsheet_with_distillation_solution_testing.ipynb:
```diff
@@ -0,0 +1,1773 @@
+   "source": [
+    "## Remark\n",
+    "\n",
+    "Currently, the `SequentialDecomposition()` tool which we will later be using to initialize the flowsheet does not support distillation column. So, we will first simulate the flowsheet without the distillation column. After it converges, we will then add the distillation column, initialize it, and then simulate the entire flowsheet."
```
I have Incorporated the suggested changes.

radhakrishnatg commented on src/Examples/Flowsheets/HDA_flowsheet_with_distillation_solution_testing.ipynb:
```diff
@@ -0,0 +1,1773 @@
+    "- ConcreteModel (to create the concrete model object)\n",
+    "- Expression (to evaluate values as a function of variables defined in the model)\n",
+    "- Objective (to define an objective function for optimization)\n",
+    "- SolverFactory (to solve the problem)\n",
```
Switched to get_solver method.

radhakrishnatg commented:
@andrewlee94 Thank you for reviewing it! I have addressed all the comments, but one. I could not figure out how to import the hda_VLE.py file from the Tutorials section. Could you please tell me how to do that? Also, please let me know if you have any other changes on your mind.

andrewlee94 commented:
@radhakrishnatg You will need to use relative imports (which can get messy in notebooks)Have a look at https://stackoverflow.com/questions/34478398/import-local-function-from-a-module-housed-in-another-directory-with-relative-im

radhakrishnatg commented:
@andrewlee94 Thanks for the link! I have now removed the duplicate file. Please let me know if you have anymore changes. Thank you!

andrewlee94 reviewed (approved):
Looks good.
